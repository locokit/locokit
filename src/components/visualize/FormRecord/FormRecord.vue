<template>
  <lck-form
    :canSubmit="completeForm"
    :displayCancelButton="false"
    :fullWidthButton="true"
    :reset="resetForm"
    :submitting="submitting.inProgress"
    @reset-form="reset=false"
    @submit="onSubmitNewRow"
  >
    <lck-data-detail
      v-if="definition"
      v-bind="$attrs"
      :definition="definition"
      mode="creation"
      :row="newRow"
      title=""
      @update-row="onUpdateRow"
      @download-attachment="$listeners['download-attachment']"
      @update-suggestions="$listeners['update-suggestions']"
      @upload-files="onUploadFiles"
    />
  </lck-form>
</template>

<script lang="ts">
import Vue from 'vue'

import DataDetail from '@/components/store/DataDetail/DataDetail.vue'
import Form from '@/components/ui/Form/Form.vue'

import { EmittedBlockEvent, LckTableRow, LckTableRowData, LckTableView, LckTableViewColumn, Submitting } from '@/services/lck-api/definitions'
import { objectFromArray } from '@/services/lck-utils/arrays'
import { getColumnDisplayValue } from '@/services/lck-utils/columns'
import { FormRecordSettings } from '@locokit/lck-glossary'
import CommunicatingBlock from '../Block/CommunicatingBlock'
import eventHub from '@/services/lck-event-hub/eventHub'

export default Vue.extend({
  name: 'FormRecord',
  mixins: [CommunicatingBlock],
  components: {
    'lck-data-detail': DataDetail,
    'lck-form': Form
  },
  props: {
    id: {
      type: String
    },
    settings: {
      type: Object as Vue.PropType<FormRecordSettings>
    },
    definition: {
      type: Object as Vue.PropType<LckTableView>,
      default: () => ({
        columns: []
      })
    },
    submitting: {
      type: Object as Vue.PropType<Submitting>,
      default: () => ({
        inProgress: false
      })
    }
  },
  data () {
    return {
      newRow: {
        id: '',
        text: '',
        data: {}
      } as LckTableRow,
      resetForm: false
    }
  },
  computed: {
    columnsObject () {
      // Get an object containing the columns as values and their ids as keys
      if (!this.definition || !this.definition.columns) return {}
      return objectFromArray<LckTableViewColumn>(this.definition.columns, 'id')
    },
    requiredColumnsIds (): Set<string> {
      // Get the required columns ids
      const requiredColumnsIds: Set<string> = new Set()
      if (this.definition?.columns) {
        this.definition.columns.forEach(column => {
          if (column.required) requiredColumnsIds.add(column.id)
        })
      }
      return requiredColumnsIds
    },
    completeForm () {
      // Return true if all required fields are defined and not null
      for (const field of this.requiredColumnsIds) {
        if (this.newRow.data[field] === null) {
          return false
        }
      }
      return true
    }
  },
  methods: {
    onUpdateRow ({ columnId, newValue }: { columnId: string; newValue: LckTableRowData }) {
      // Update the data
      this.$set(this.newRow.data, columnId, newValue)
      // Throw updated events to other blocks if it's specified
      const triggerEvent = this.settings.triggerEvents?.find(event =>
        event.type === 'update' &&
        event.field === columnId
      )
      if (triggerEvent) {
        eventHub.$emit(triggerEvent.name, {
          originalValue: newValue,
          displayedValue: getColumnDisplayValue(this.columnsObject[columnId], newValue, true)
        } as EmittedBlockEvent)
      }
    },
    onSubmitNewRow () {
      this.$emit('create-row', this.newRow)
    },
    resetNewRow () {
      this.newRow = {
        id: '',
        text: '',
        data: {}
      }
      this.resetForm = true
    },
    onUploadFiles (event: { rowId: string; columnId: string; fileList: File[]}) {
      this.$emit('upload-files', {
        ...event,
        newRow: this.newRow
      })
    },
    onSelectBlockEvent (columnId: string | undefined, { originalValue }: EmittedBlockEvent) {
      // Catch an event coming from another block to affect the value to a specific field
      if (columnId && originalValue !== undefined) {
        const { data } = this.newRow
        data[columnId] = originalValue
        this.newRow.data = {
          ...data
        }
      }
    }
  },
  watch: {
    submitting (submittingValue: Submitting) {
      if (!submittingValue.inProgress && !submittingValue.errors) {
        // When submitting without any error
        this.resetNewRow()
        // Throw an event saying that the form has been submitting
        const triggerEvent = this.settings.triggerEvents?.find(event => event.type === 'submit')
        if (triggerEvent) {
          eventHub.$emit(triggerEvent.name, {} as EmittedBlockEvent)
        }
      }
    }
  }
})
</script>
