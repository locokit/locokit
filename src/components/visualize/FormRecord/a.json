{"sources": [{"id": "464d6769-498d-4e2f-b0f8-dd1599a8565c", "popup": true, "popupSettings": { "title": "Informations", "contentFields": [{ "field": "32ca6a9f-b27a-4f0e-bb5e-3cc554fd4a55"}]}}, {"id": "5fb96ecf-f8f7-4476-9292-a4bbfa0087f9", "triggerEvents": ["click"]}]}