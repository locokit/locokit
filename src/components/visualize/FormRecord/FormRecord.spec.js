/* eslint-disable @typescript-eslint/camelcase */
import { BLOCK_TYPE, COLUMN_TYPE } from '@locokit/lck-glossary'
import { shallowMount } from '@vue/test-utils'

import FormRecord from '@/components/visualize/FormRecord/FormRecord.vue'

const mockDefinitionsWithRequiredColumns = {
  id: 'table_view_1',
  columns: [
    {
      text: 'Description',
      id: 'string_1_column',
      settings: {},
      table_id: 'table_1',
      column_type_id: COLUMN_TYPE.STRING,
      editable: false,
      position: 1,
      transmitted: false,
      default: '',
      displayed: true,
      required: true,
      sort: 'DESC',
      table_column_id: '',
      table_view_id: 'table_view_1',
      style: {}
    },
    {
      text: 'Number',
      id: 'number_1_column',
      settings: {},
      table_id: 'table_1',
      column_type_id: COLUMN_TYPE.NUMBER,
      editable: false,
      position: 1,
      transmitted: false,
      default: '',
      displayed: true,
      required: true,
      sort: 'DESC',
      table_column_id: '',
      table_view_id: 'table_view_1',
      style: {}
    },
    {
      text: 'Boolean',
      id: 'boolean_1_column',
      settings: {},
      table_id: 'table_1',
      column_type_id: COLUMN_TYPE.BOOLEAN,
      editable: false,
      position: 1,
      transmitted: false,
      default: '',
      displayed: true,
      required: true,
      sort: 'DESC',
      table_column_id: '',
      table_view_id: 'table_view_1',
      style: {}
    }
  ]
}

const mockDefinitionsWithoutRequiredColumns = {
  id: 'table_view_2',
  columns: [
    {
      text: 'Description',
      id: 'string_2_column',
      settings: {},
      table_id: 'table_1',
      column_type_id: COLUMN_TYPE.STRING,
      editable: false,
      position: 1,
      transmitted: false,
      default: '',
      displayed: true,
      required: false,
      sort: 'DESC',
      table_column_id: '',
      table_view_id: 'table_view_2',
      style: {}
    },
    {
      text: 'Number',
      id: 'number_2_column',
      settings: {},
      table_id: 'table_1',
      column_type_id: COLUMN_TYPE.NUMBER,
      editable: false,
      position: 1,
      transmitted: false,
      default: '',
      displayed: true,
      required: false,
      sort: 'DESC',
      table_column_id: '',
      table_view_id: 'table_view_2',
      style: {}
    }
  ]
}

describe('FormRecord', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallowMount(FormRecord, {
      attrs: {
        workspaceId: 'workspace_1'
      },
      listeners: {
        'download-attachment': () => ({}),
        'update-suggestions': () => ({}),
        'upload-files': () => ({})
      }
    })
    expect(wrapper.vm.requiredColumnsIds.size).toBe(0)
  })

  describe('Computed properties', () => {
    describe('requiredColumnsIds', () => {
      it('Returns a set of the required columns ids if there are several ones', async () => {
        await wrapper.setProps({
          definition: mockDefinitionsWithRequiredColumns,
          settings: {
            id: mockDefinitionsWithRequiredColumns.id
          }
        })
        expect(wrapper.vm.requiredColumnsIds.size).toBe(3)
        expect(wrapper.vm.requiredColumnsIds.has('string_1_column')).toBe(true)
        expect(wrapper.vm.requiredColumnsIds.has('boolean_1_column')).toBe(true)
        expect(wrapper.vm.requiredColumnsIds.has('number_1_column')).toBe(true)
      })
      it('Returns an empty set if there is no required column', async () => {
        await wrapper.setProps({
          definition: mockDefinitionsWithoutRequiredColumns,
          settings: {
            id: mockDefinitionsWithoutRequiredColumns.id
          }
        })
        expect(wrapper.vm.requiredColumnsIds.size).toBe(0)
      })
      it('Returns an empty set if there is no definition', async () => {
        await wrapper.setProps({
          definition: null
        })
        expect(wrapper.vm.requiredColumnsIds.size).toBe(0)
      })
    })
    describe('completeForm', () => {
      it('Return false if at least one required field is complete', async () => {
        await wrapper.setProps({
          definition: mockDefinitionsWithRequiredColumns
        })
        wrapper.setData({
          newRow: {
            data: {
              string_1_column: 'Data',
              number_1_column: 10
            }
          }
        })
        expect(wrapper.vm.completeForm).toBe(false)
      })
      it('Return true if all required fields are complete', async () => {
        await wrapper.setProps({
          definition: mockDefinitionsWithRequiredColumns
        })
        wrapper.setData({
          newRow: {
            data: {
              string_1_column: 'Data',
              number_1_column: 10,
              boolean_1_column: false
            }
          }
        })
        expect(wrapper.vm.completeForm).toBe(true)
      })
    })
  })
})
