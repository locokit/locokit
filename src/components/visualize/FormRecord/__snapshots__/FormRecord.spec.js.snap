// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRecord could have required columns disable the submit button if they are not set 1`] = `
<span workspaceid="workspace_1" crudmode="true"><form class="p-fluid"><div class="p-fluid"><!----> <div class="p-field"><span style="position: relative;"><div class="p-d-flex" style="position: relative;"><!----> <label for="string_1_column" class="lck-color-primary untouched pristine pending required">
          Description
        </label> <span class="field-required">
          *
        </span>
<!---->
</div>
<div class="form-field-editable untouched pristine pending required"><input class="p-inputtext p-component p-filled" id="string_1_column" rows="7"> <span class="cell-state"></span></div> <span class="untouched pristine pending required"></span></span></div>
<div class="p-field"><span style="position: relative;"><div class="p-d-flex" style="position: relative;"><!----> <label for="number_1_column" class="lck-color-primary untouched pristine pending">
          Number
        </label> <!----> <!----></div> <div class="form-field-editable untouched pristine pending"><span class="p-inputnumber p-component p-inputwrapper-filled"><input class="p-inputtext p-component p-filled p-inputnumber-input" id="number_1_column" rows="7"> <!----> <!----> <!----></span> <span class="cell-state"></span></div> <span class="untouched pristine pending"></span></span></div>
<div class="p-field"><span style="position: relative;"><div class="p-d-flex" style="position: relative;"><div class="p-checkbox p-component p-checkbox-checked"><div class="p-hidden-accessible"><input type="checkbox" id="boolean_1_column" value=""></div> <div role="checkbox" class="p-checkbox-box p-highlight" aria-checked="true"><span class="p-checkbox-icon pi pi-check"></span></div>
</div> <label for="boolean_1_column" class="lck-color-primary untouched pristine pending">
  Boolean
</label>
<!----> <span class="cell-state"></span></div>
<div class="form-field-editable untouched pristine pending">
  <!---->
  <!---->
</div> <span class="untouched pristine pending"></span></span></div>
</div>
<div class="lck-form-footer">
  <!----> <button type="submit" class="p-button p-component full-width-button" disabled="disabled"><span class="pi pi-save p-button-icon p-button-icon-left"></span> <span class="p-button-label">form.save</span>
    <!----></button></div>
</form>
</span>
`;

exports[`FormRecord could have required columns disable the submit button if they are not valid 1`] = `
<span workspaceid="workspace_1" crudmode="true"><form class="p-fluid"><div class="p-fluid"><!----> <div class="p-field"><span style="position: relative;"><div class="p-d-flex" style="position: relative;"><!----> <label for="string_1_column" class="lck-color-primary untouched pristine required">
          Description
        </label> <span class="field-required">
          *
        </span>
<!---->
</div>
<div class="form-field-editable untouched pristine required"><input class="p-inputtext p-component" id="string_1_column" rows="7"> <span class="cell-state"></span></div> <span class="untouched pristine required"></span></span></div>
<div class="p-field"><span style="position: relative;"><div class="p-d-flex" style="position: relative;"><!----> <label for="number_1_column" class="lck-color-primary untouched pristine">
          Number
        </label> <!----> <!----></div> <div class="form-field-editable untouched pristine"><span class="p-inputnumber p-component"><input class="p-inputtext p-component p-inputnumber-input" id="number_1_column" rows="7"> <!----> <!----> <!----></span> <span class="cell-state"></span></div> <span class="untouched pristine"></span></span></div>
<div class="p-field"><span style="position: relative;"><div class="p-d-flex" style="position: relative;"><div class="p-checkbox p-component"><div class="p-hidden-accessible"><input type="checkbox" id="boolean_1_column" value=""></div> <div role="checkbox" class="p-checkbox-box"><span class="p-checkbox-icon"></span></div>
</div> <label for="boolean_1_column" class="lck-color-primary untouched pristine">
  Boolean
</label>
<!----> <span class="cell-state"></span></div>
<div class="form-field-editable untouched pristine">
  <!---->
  <!---->
</div> <span class="untouched pristine"></span></span></div>
</div>
<div class="lck-form-footer">
  <!----> <button type="submit" class="p-button p-component full-width-button" disabled="disabled"><span class="pi pi-save p-button-icon p-button-icon-left"></span> <span class="p-button-label">form.save</span>
    <!----></button></div>
</form>
</span>
`;

exports[`FormRecord could have required columns enable the submit button if all fields are valid and at least one field is updated 1`] = `
<span workspaceid="workspace_1" crudmode="true"><form class="p-fluid"><div class="p-fluid"><!----> <div class="p-field"><span style="position: relative;"><div class="p-d-flex" style="position: relative;"><!----> <label for="string_1_column" class="lck-color-primary untouched dirty valid validated required changed passed">
          Description
        </label> <span class="field-required">
          *
        </span>
<!---->
</div>
<div class="form-field-editable untouched dirty valid validated required changed passed"><input class="p-inputtext p-component p-filled" id="string_1_column" rows="7"> <span class="cell-state"></span></div> <span class="untouched dirty valid validated required changed passed"></span></span></div>
</div>
<div class="lck-form-footer">
  <!----> <button type="submit" class="p-button p-component full-width-button"><span class="pi pi-save p-button-icon p-button-icon-left"></span> <span class="p-button-label">form.save</span>
    <!----></button></div>
</form>
</span>
`;

exports[`FormRecord could have required columns enable the submit button if all fields are valid and at least one field is updated 2`] = `
<span workspaceid="workspace_1" crudmode="true"><form class="p-fluid"><div class="p-fluid"><!----> <div class="p-field"><span style="position: relative;"><div class="p-d-flex" style="position: relative;"><!----> <label for="string_1_column" class="lck-color-primary untouched dirty valid validated required changed passed">
          Description
        </label> <span class="field-required">
          *
        </span>
<!---->
</div>
<div class="form-field-editable untouched dirty valid validated required changed passed"><input class="p-inputtext p-component p-filled" id="string_1_column" rows="7"> <span class="cell-state"></span></div> <span class="untouched dirty valid validated required changed passed"></span></span></div>
</div>
<div class="lck-form-footer">
  <!----> <button type="submit" class="p-button p-component full-width-button"><span class="pi pi-save p-button-icon p-button-icon-left"></span> <span class="p-button-label">form.save</span>
    <!----></button></div>
</form>
</span>
`;

exports[`FormRecord could have required columns enable the submit button if no required fields are defined and at least one field is updated 1`] = `
<span workspaceid="workspace_1" crudmode="true"><form class="p-fluid"><div class="p-fluid"><!----> <div class="p-field"><span style="position: relative;"><div class="p-d-flex" style="position: relative;"><!----> <label for="string_2_column" class="lck-color-primary untouched dirty valid validated changed passed">
          Description
        </label> <!----> <!----></div> <div class="form-field-editable untouched dirty valid validated changed passed"><input class="p-inputtext p-component p-filled" id="string_2_column" rows="7"> <span class="cell-state"></span></div> <span class="untouched dirty valid validated changed passed"></span></span></div>
</div>
<div class="lck-form-footer">
  <!----> <button type="submit" class="p-button p-component full-width-button"><span class="pi pi-save p-button-icon p-button-icon-left"></span> <span class="p-button-label">form.save</span>
    <!----></button></div>
</form>
</span>
`;
